packages/squiggle-lang/src/js/math.ts

```diff
@@ -0,0 +1,113 @@
+type SampleArgs = {
+  probs: number[];
+  size: number;
+};
+
+/** Searches for a first element in `arr` greater than `value` */
+function binsearchFirstGreater(arr: number[], value: number): number {
+  let s = 0
+  let e = arr.length - 1;
+  while ((e - s) > 1) {
+    const i = Math.ceil((s + e) / 2)
+    if (arr[i] < value) {
+      s = i + 1
+    } else {
+      e = i
+    }
+  }
+  return value < arr[s] ? s : e
+}
+
+export function randomSample(dist: number[], args: SampleArgs): number[] {
+  const { probs, size } = args;
+  let sample: number[] = Array(size);
+
+  let accum = 0
+  let probPrefixSums = [];
+  for (const f of probs) {
+    probPrefixSums.push(accum += f)
+  }
+  let sum = probPrefixSums[probPrefixSums.length - 1]
+
+  for (let index = 0; index < size; index++) {
+    let selection = binsearchFirstGreater(probPrefixSums, Math.random() * sum)
+    sample[index] = dist[selection];
+  }
+  return sample;
+}
+
+// Should be replaced by a lookup table if we ever use it too much
+export function factorial(n: number): number {
+  let fct = 1;
+  while (n > 0) {
+    fct *= n--
+  }
+  return fct;
+}
+
+type Distribution<P> = {
+  cdf(args: P, x: number): number;
+  mean(args: P): number;
+  stdev(args: P): number;
+  variance(args: P): number;
+  quantile(args: P, p: number): number;
+};
+
+type ContinuousProbabilityDistribution<P> = Distribution<P> & {
+  pdf(args: P, x: number): number;
+};
+
+type DiscreteProbabilityDistribution<P> = Distribution<P> & {
+  pmf(args: P, x: number): number;
+};
+
+let bernoulli: DiscreteProbabilityDistribution<number> = {
+  pmf: (p, x) => (x === 0 ? 1 - p : p),
+  cdf: (p, x) => (x < 0 ? 0 : x >= 1 ? 1 : 1 - p),
+  mean: (p) => p,
+  stdev: (p) => Math.sqrt(p * (1 - p)),
+  quantile: (p, prob) => (prob <= 1 - p ? 0 : 1),
+  variance: (p) => p * (1 - p),
+};
+
+function sq(n: number): number { return n * n }
+
+let logistic: ContinuousProbabilityDistribution<{ mu: number; s: number }> = {
```

Comment:
I imagine later we might just want a ``logistic`` type with `mu` and `s` in it, instead of manually passing each.

Reply:
I've created those types mostly so I don't make any mistakes at least there)

It sure would be nice to have proper distribution classes, yes)